### PR TITLE
Auto-fix corrupt JPEGs

### LINKED ISSUE #916: premature end of JPEG images
## ❔Question
`Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
      1/99     2.87G   0.05456   0.04197         0   0.09652        10       640: 100% 157/157 [00:52<00:00,  2.98it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95:   0% 0/157 [00:00<?, ?it/s]Premature end of JPEG file
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100% 157/157 [00:19<00:00,  8.21it/s]
                 all     2.5e+03       1e+04       0.362       0.777       0.684       0.338`

It shows premature end of JPEG images during validation, what leads to this?

### LINKED ISSUE #2451: corrupt JPEG data
Excuse me.When i was training my data, the cmd would show this problem: 'corrupt JPEG data:Premature end of data segment'. Would it influence the result? My  train data adds some background images.

### LINKED ISSUE #3450: Corrupt JPEG data
## ❔Question
If I test the image and it show the message.
How can I debug it? 
Corrupt JPEG data: 1 extraneous bytes before marker 0xd9
Corrupt JPEG data: 1 extraneous bytes before marker 0xd9

### LINKED ISSUE #3731: Corrupted JPEG,I can't solved.
## ❔Question
Why is my image or label corrupted? A lot of pictures are corrupted, so I checked some of them and opened the pictures and the corresponding label. There is no problem.
Can you give me some help?
I tried to delete cacha and retrain
I looked at utiles / datasets.py, but I didn't know how to debug it.

## Additional context
`train: WARNING: Ignoring corrupted image and/or label voc/images/9556041603720_camera3-8.jpg: corrupted JPEG
train: WARNING: Ignoring corrupted image and/or label voc/images/9556041603720_camera3-9.jpg: corrupted JPEG
train: Scanning 'voc/train' images and labels...29281 found, 0 missing, 0 empty, 46085 corrupted: 100%|█| 75366/75366 [00:08<00
train: New cache created: voc/train.cache

val: WARNING: Ignoring corrupted image and/or label voc/images/9556041603720_camera2-25.jpg: corrupted JPEG
val: WARNING: Ignoring corrupted image and/or label voc/images/6907777808584-back_camera2-20.jpg: corrupted JPEG
val: Scanning 'voc/val' images and labels...333 found, 0 missing, 0 empty, 505 corrupted: 100%|███████| 838/838 [00:00<00:00, 1
val: New cache created: voc/val.cache
Plotting labels...

autoanchor: Analyzing anchors... anchors/target = 3.51, Best Possible Recall (BPR) = 1.0000
Image sizes 1280 train, 1280 test
Using 4 dataloader workers
Logging results to runs/train/RPC
Starting training for 600 epochs...

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
     0/599     20.3G   0.06935    0.1463    0.1145    0.3301        72      1280:  37%|█████         | 2673/7321 [37:15<1:03:24     0/599     20.3G   0.06935    0.1463    0.1145    0.3301        72      1280:  37%|█████         | 2673/7321 [37:15<1:04:47
`

### PR DESCRIPTION
**Pull request by glenn-jocher**
This PR automatically re-saves corrupt JPEGs and trains with the resaved images.

Current behavior is to exclude corrupt JPEGs from training with a warning to the user, but many users have been complaining about large parts of their dataset being excluded from training.

**WARNING**: this will overwrite the existing corrupt JPEGs in a dataset and replace them with correct JPEGs, though the filesize may increase and the image contents may not be exactly the same due to lossy JPEG compression schemes. Results may vary by JPEG decoder and hardware.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements for image and label processing in YOLOv5.

### 📊 Key Changes
- Removed unnecessary stacking of image shapes, simplifying the shape checking logic.
- Added a comment spacing for better code readability.
- Included initialization of a `msg` variable and a `segments` list to handle image and label verification messages and segment data respectively.
- Implemented a corruption check and auto-repair for JPEG images, ensuring they end with the proper byte sequence.

### 🎯 Purpose & Impact
- 🔍 **Image Shape Checking**: Streamlining this process helps maintain efficiency in handling input images.
- 📝 **Readability Improvement**: Better spacing around comments enhances readability for developers.
- 🛠️ **Verification Robustness**: The added variables `msg` and `segments` along with JPEG repair functionality means better fault tolerance and informative messages when verifying images and labels, improving overall data integrity.
- 💼 **Potential Impact**: These changes can lead to fewer errors during training/validation with varying quality of datasets, consequently improving the user experience and reliability of YOLOv5.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Allclose fails, but visually the images appear identical.

```python
import cv2
import matplotlib.pyplot as plt
import numpy as np
from PIL import Image

fp = '000000000034.jpg'  # original image file path
fp_pil = fp + '.PIL.jpg'
fp_cv2 = fp + '.cv2.jpg'

# Read and write cv2 and PIL JPGs
Image.open(fp).save(fp_pil)
cv2.imwrite(fp_cv2, cv2.imread(fp))

# Read new JPGs and compare
im = cv2.imread(fp)
im_pil = cv2.imread(fp_pil)
im_cv2 = cv2.imread(fp_cv2)
dp = (im - im_pil).ravel()
dc = (im - im_cv2).ravel()
print(np.allclose(im, im_pil))
print(np.allclose(im, im_cv2))

# Plot
fig, ax = plt.subplots(1, 2, figsize=(8, 4), tight_layout=True)
ax[0].hist(dp, 255)
ax[1].hist(dc, 255)
plt.savefig('results.jpg')
```

results

Related: https://stackoverflow.com/questions/54610705/copied-image-saved-with-different-pixels-to-original-with-pil

**Review by Poulinakis-Konstantinos — APPROVED** [review]
Changes are approved. Logic is correct. 

Maybe you could give a better hint to the user as to why it's being re-saved,
"WARNING: Re-saving corrupted JPEG {im_file} to restore it"

**Review by Poulinakis-Konstantinos — DISMISSED** [review]

**Comment by glenn-jocher** [comment]
@Poulinakis-Konstantinos PR is merged after clarifying warning message. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Comment by Poulinakis-Konstantinos** [comment]
@glenn-jocher It was my pleasure to help 😃  ! 

Thank you for all the valuable contributions ! 🚀

**Comment by Djordi97** [comment]
Hi @glenn-jocher. It appears as if this fix does not fully work for the SKU-110K dataset. When running the command `python train.py --data SKU-110K.yaml --cfg yolov5n.yaml` I get the following output from scanning of the dataset:

Screenshot 2022-06-17 at 14 05 44

The scan detected 3 corrupted images in the training set. However, during the training phase it turns out that there are many more corrupt files:

Screenshot 2022-06-17 at 14 08 50

Any idea on how to fix this? Thanks a lot in advance!

**Comment by glenn-jocher** [comment]
@Djordi97 python scans can only fix images with errors reported in python. Lower level decompression algorithms are reporting the errors shown in your training.

**Comment by Djordi97** [comment]
@glenn-jocher ahh I see, thanks for the explanation. I think for the SKU dataset this concerns ~100 images. Do you know if there is any way to resolve this issue? And is it safe to include these images during training or should I try to filter them out? Perhaps its possible to improve (e.g. filter, restore) this dataset programatically since it is included in the repository and can prevent others from running into the same issue?

**Comment by glenn-jocher** [comment]
@Djordi97 SKU-110K is downloaded from other servers that are not our own, and is already included in publications with benchmarks etc, so despite there being defects to compare apples to apples you'd want to train on the same exact dataset. 

If you wanted to fix the JPEGs you could probably do so by loading and saving each one with cv2, though you should be very careful for saving/compression artifacts.

**Comment by Djordi97** [comment]
@glenn-jocher yeah that's a good point indeed. For objective scientific comparison you would need to train on the same dataset, although my use case is slightly different. Think the authors could have spend a bit more time on checking the quality of the dataset (believe it is crowdsourced). 

Thanks a lot for the pointer, I will have a look at it!

**Comment by glenn-jocher** [comment]
@Djordi97 yes, unfortunately it's a fact of life in the ML space that many datasets have slight issues here and there, even the ones from reputable sources like COCO, ImageNet etc. With thousands of images you should expect a few problems and make your workflow robust to this, which is what we've tried to do with YOLOv5.
